### PR TITLE
COS: Added support for Noncurrent version expiration, EODM & abort incomplete MPU

### PR DESCRIPTION
**Pull request by Priyaranimishra**
COS Objectversioning ph2 & abort incomplete MPU:----

-  Incomplete multi-part upload cleanup. 
-  Object Expiration for Versioned Buckets New Lifecycle Functionality
-  Non-Current Version Expiration Lifecycle
-  Expiration Lifecycle behavior on versioning-enabled buckets
-  Expired Object Delete Marker Lifecycle

### CODE REVIEW

**Comment by Priyaranimishra** [comment]
Here is the test results:-

=== RUN   TestAccIBMCosBucket_noncurrentversion
--- PASS: TestAccIBMCosBucket_noncurrentversion (107.00s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	109.462s

=== RUN   TestAccIBMCosBucket_AbortIncompeleteMPU
--- PASS: TestAccIBMCosBucket_AbortIncompeleteMPU (105.43s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	108.222s

=== RUN   TestAccIBMCosBucket_Expiredays
--- PASS: TestAccIBMCosBucket_Expiredays (163.35s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	165.730s

=== RUN   TestAccIBMCosBucket_Expiredate
--- PASS: TestAccIBMCosBucket_Expiredate (126.77s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	130.008s

=== RUN   TestAccIBMCosBucket_Expireddeletemarker
--- PASS: TestAccIBMCosBucket_Expireddeletemarker (74.71s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	77.028s

File: ibm/resource_ibm_cos_bucket.go [hunk 1]
```diff
@@ -227,7 +260,7 @@ func resourceIBMCOSBucket() *schema.Resource {
 			"expire_rule": {
 				Type:        schema.TypeList,
 				Optional:    true,
-				MaxItems:    1000,
+				MaxItems:    1,
```
Comment (hkantare) [inline ibm/resource_ibm_cos_bucket.go:263]: you can create multiple expiry rules?
why the number is changed from 1000 to 1

File: ibm/resource_ibm_cos_bucket.go [hunk 2]
```diff
@@ -296,11 +343,10 @@ func resourceIBMCOSBucket() *schema.Resource {
...
 			"object_versioning": {
-				Type:          schema.TypeList,
-				Optional:      true,
-				MaxItems:      1,
-				ConflictsWith: []string{"retention_rule", "expire_rule"},
-				Description:   "Protect objects from accidental deletion or overwrites. Versioning allows you to keep multiple versions of an object protecting from unintentional data loss.",
+				Type:        schema.TypeList,
+				Optional:    true,
+				MaxItems:    1,
+				Description: "Protect objects from accidental deletion or overwrites. Versioning allows you to keep multiple versions of an object protecting from unintentional data loss.",
```
Comment (hkantare) [inline ibm/resource_ibm_cos_bucket.go:349]: why is conflict removed here we can proivide object_versioning along with retention_rule and expire_rule

File: website/docs/d/cos_bucket.html.markdown [hunk 3]
```diff
@@ -47,7 +47,14 @@ Review the argument references that you can specify for your data source.
 
 ## Attribute reference
 In addition to all argument reference list, you can access the following attribute references after your data source is created. 
-
+- `crn` - (String) The CRN of the bucket.
+- `cross_region_location` - (String) The location to create a cross-regional bucket.
+- `single_site_location` - (String) The location to create a single site bucket.
```
Comment (hkantare) [inline website/docs/d/cos_bucket.html.markdown:52]: why is the order changed they were arranged in alphabetical order

File: website/docs/r/cos_bucket.html.markdown [hunk 4]
```diff
@@ -221,26 +284,48 @@ Review the argument references that you can specify for your resource.
...
-- `expire_rule` - (Required, List) Nested expire_rule block has following structure.
+    **Note:** 
+    - Archive is available in certain regions only. For more information, see [Integrated Services](https://cloud.ibm.com/docs/cloud-object-storage/basics?topic=cloud-object-storage-service-availability).
+    - Restoring object once archive is not supported yet.
+- `expire_rule` - (Required, List) An expiration rule deletes objects after a defined period (from the object creation date). see [lifecycle actions](https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-versioning). Nested expire_rule block has following structure.
   
   Nested scheme for `expire_rule`:
   - `rule_id` -  (Optional, Computed, String) Unique ID for the rule. Expire rules allow you to set a specific time frame after which objects are deleted.
   - `enable` - (Required, Bool) Specifies expire rule status either `enable` or `disable` for a bucket.
-  - `days` - (Required, String) Specifies the number of days when the specific rule action takes effect.
```
Comment (hkantare) [inline website/docs/r/cos_bucket.html.markdown:235]: Arrange the nested structure inner attributes in alphabetical order for all rules

**Comment by hkantare** [comment]
share the test results of old testcases also

**Comment by Priyaranimishra** [comment]
here is the rest test cases result except TestAccIBMCosBucket_ActivityTracker_Monitor

=== RUN   TestAccIBMCosBucket_Archive_Expiration
--- PASS: TestAccIBMCosBucket_Archive_Expiration (134.26s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	136.677s

=== RUN   TestAccIBMCosBucket_Archive_Expiration
--- PASS: TestAccIBMCosBucket_Archive_Expiration (140.09s)
=== RUN   TestAccIBMCosBucket_Archive
--- PASS: TestAccIBMCosBucket_Archive (133.38s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	276.198s


=== RUN   TestAccIBMCosBucket_Retention
--- PASS: TestAccIBMCosBucket_Retention (82.45s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	85.163s
(base) Priyas-MBP:terraform-provider-ibm priyarani$

=== RUN   TestAccIBMCosBucket_Object_Versioning
--- PASS: TestAccIBMCosBucket_Object_Versioning (76.08s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	78.983s


=== RUN   TestAccIBMCosBucket_Hard_Quota
--- PASS: TestAccIBMCosBucket_Hard_Quota (70.33s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	72.706s

=== RUN   TestAccIBMCosBucket_Smart_Type
--- PASS: TestAccIBMCosBucket_Smart_Type (96.37s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	109.880s

=== RUN   TestAccIBMCosBucket_import
--- PASS: TestAccIBMCosBucket_import (80.09s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	82.808s\\

=== RUN   TestAccIBMCosBucket_Expireddeletemarker
--- PASS: TestAccIBMCosBucket_Expireddeletemarker (70.32s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	72.774s

File: ibm/resource_ibm_cos_bucket.go (hunk 1 above)
Comment (Priyaranimishra) [reply ibm/resource_ibm_cos_bucket.go:263]: Max item should be 1 only . Once we change to 1000, it failed with below error in testing
```
resource_ibm_cos_bucket_test.go:92: Step 1/2 error: Error running pre-apply refresh: exit status 1
        2021/11/24 20:46:00 [DEBUG] Using modified User-Agent: Terraform/0.12.28 HashiCorp-terraform-exec/0.14.0

        Error: InternalValidate

        Internal validation of the provider failed! This is always a bug
        with the provider itself, and not a user issue. Please report
        this bug:

        1 error occurred:
        	* resource ibm_cos_bucket: ConflictsWith: expired_object_delete_marker
        configuration block reference (expire_rule.0.date) can only be used with
        TypeList and MaxItems: 1 configuration blocks
```

**Comment by Priyaranimishra** [comment]
Test result after making changes with Maxitems to 1000

--- PASS: TestAccIBMCosBucket_Expiredays (149.91s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	152.856s


--- PASS: TestAccIBMCosBucket_Expiredate (272.23s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	281.453s

**Pull request merged**